### PR TITLE
Allow replayed producers to be restarted after termination

### PR DESCRIPTION
**Pull request by sharplet**
Related to #97. I've been to get together a proof of concept of the API described in https://github.com/ReactiveCocoa/ReactiveSwift/issues/97#issuecomment-263199750, but running into subtle issues with the lifetime semantics of composed Properties (which use `replayLazily` internally).

Currently, the start handler goes out of scope the first time it is started, which means no strong reference is kept to the inner producer after starting. In order to allow the replayed producer to be restarted, we need to keep a reference to it (see `startHandler`), but keeping this reference around breaks a couple of tests verifying when a composed property sends a `completed` event.

It looks like `Property.init(unsafeProducer:)` is implicitly relying on the fact that `replayLazily` discards its reference to the inner producer after the first time it's started. I've tried a few different things, but I haven't been able to work out an alternate implementation of the initialiser that doesn't rely on that fact.

`Property` lifetime management still feels pretty hard! I _think_ in part because `SignalProducer` is a struct, I'm finding it difficult to reason about who owns a reference to what, and why the test is failing. I'd really appreciate some help working out what's going on here.

### CODE REVIEW

**Comment by sharplet** [comment]
I've pushed d7b6686bba9a5f939ee926e3f7e82e7789630344 to show where I'm currently going with the replay cache policy idea. Once CI completes, it should hopefully show that 6cab6a3bde2c92a8fd589828bd765785e03594ab is the minimal change that breaks the composed `Property` lifetime behaviours, and d7b6686bba9a5f939ee926e3f7e82e7789630344 adds support for retrying after failure but without breaking the existing tests for `replayLazily()`. This is still very much a work in progress, as I got fairly set back by the property lifetime stuff and haven't tested the new stuff yet.

File: Sources/SignalProducer.swift [hunk 1]
```diff
@@ -1866,9 +1894,11 @@ extension SignalProducerProtocol {
 					}
 
 					// Start the underlying producer if it has never been started.
-					start.swap(nil)?()
+					if start.swap(false) {
+						startHandler()
+					}
 
-					// Terminate the replay loop.
+					// Terminate the observation loop.
```
Comment (andersio) [inline Sources/SignalProducer.swift]: I don't feel the need to rename it. 🤔

File: Sources/SignalProducer.swift [hunk 2]
```diff
@@ -1830,17 +1833,43 @@ extension SignalProducerProtocol {
 		let lifetime = Lifetime(lifetimeToken)
 
 		let state = Atomic(ReplayState<Value, Error>(upTo: capacity))
+		let start = Atomic(true)
 
-		let start: Atomic<(() -> Void)?> = Atomic {
-			// Start the underlying producer.
+		// Start the underlying producer.
+		func startHandler() {
```
Comment (andersio) [inline Sources/SignalProducer.swift]: `replayLazily` retains `self` only when it is not ever started. Once it is started, it throws the start handler (which retained `self`) away. This is why there is a `start` of type `Atomic<(() -> Void)?>`. The behaviour — at least for the default policy — should not be changed.

File: Sources/SignalProducer.swift [hunk 3]
```diff
@@ -1830,17 +1833,43 @@ extension SignalProducerProtocol {
...
-				.start { event in
-					let observers: Bag<Signal<Value, Error>.Observer>? = state.modify { state in
-						defer { state.enqueue(event) }
-						return state.observers
-					}
-					observers?.forEach { $0.action(event) }
+				.startWithSignal { events, _ in
+					let events = SignalProducer(events)
+
+					let invalidated = events
```
Comment (andersio) [inline Sources/SignalProducer.swift]: It seems all these could be flattened into one event action.

File: Sources/SignalProducer.swift (hunk 2 above)
Comment (andersio) [inline Sources/SignalProducer.swift]: I'd support an eager release of the start handler whenever it is possible though.

File: Sources/SignalProducer.swift [hunk 4]
```diff
@@ -1830,17 +1833,43 @@ extension SignalProducerProtocol {
...
 		let state = Atomic(ReplayState<Value, Error>(upTo: capacity))
+		let start = Atomic(true)
 
-		let start: Atomic<(() -> Void)?> = Atomic {
-			// Start the underlying producer.
+		// Start the underlying producer.
+		func startHandler() {
 			self
+				.materialize()
 				.take(during: lifetime)
```
Comment (andersio) [inline Sources/SignalProducer.swift]: The lifetime is what causes the failing test cases. The token tracks all outstanding producers, and breaks the observation only if there is none of them left.

**Comment by andersio** [comment]
The source of the problem is the guarantee of all properties emitting its latest value, even if it has been deinitialized.

In these failing test cases:

1. the flattened property is retained till the end of the scope.
2. As it was made to retain the upstream producer in `startHandler`, this means ultimately the outer property's producer is retained too.
3. The producer retains the `Atomic` storage, which retains the last property being sent.

This also explains why `latest` doesn't fail — the flattened property has deinitialized immediately after use.

I don't think there is any workaround.

**Comment by andersio** [comment]
I am still leaning to an overload constrained to a special error type. The PR in its current form would allow `failure` being silently absorbed by `replayLazily`, regardless of being intentional or not.

**Comment by andersio** [comment]
We might consider instead generalizing this as "autoconnect()" operator like Combine and Rx when there is an interest in expanding primitive operators.

Closing because of inactivity.

**Pull request closed without merging**